flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2134,8 +2126,77 @@ class ArrayExprLowering {
         exp.u);
   }
 
+  /// Compute the shape of a slice.
+  llvm::SmallVector<mlir::Value> computeSliceShape(mlir::Value slice) {
+    llvm::SmallVector<mlir::Value> slicedShape;
+    auto slOp = mlir::cast<fir::SliceOp>(slice.getDefiningOp());
+    auto triples = slOp.triples();
+    auto idxTy = builder.getIndexType();
+    auto loc = getLoc();
+    for (unsigned i = 0, end = triples.size(); i + 2 < end; i += 3) {
+      if (!mlir::isa_and_nonnull<fir::UndefOp>(
+              triples[i + 1].getDefiningOp())) {
+        // (..., lb:ub:step, ...) case:  extent = (ub-lb+step)/step
+        auto lb = builder.createConvert(loc, idxTy, triples[i]);
+        auto ub = builder.createConvert(loc, idxTy, triples[i + 1]);
+        auto step = builder.createConvert(loc, idxTy, triples[i + 2]);
+        auto diff = builder.create<mlir::SubIOp>(loc, ub, lb);
+        auto add = builder.create<mlir::AddIOp>(loc, diff, step);
+        slicedShape.emplace_back(
+            builder.create<mlir::SignedDivIOp>(loc, add, step));
+      }
+      // else (..., i, ...) case: drop dimension.
+    }
+    return slicedShape;
+  }
+
+  /// Get the shape from an array load.
+  llvm::SmallVector<mlir::Value> getShape(fir::ArrayLoadOp arrayLoad) {
+    if (arrayLoad.slice())
+      return computeSliceShape(arrayLoad.slice());
+    auto memref = arrayLoad.memref();
+    if (memref.getType().template isa<fir::BoxType>()) {
+      llvm::SmallVector<mlir::Value> shape;
+      Fortran::lower::readExtents(builder, getLoc(), fir::BoxValue{memref},
+                                  shape);
+      return shape;
+    }
+    auto extents = arrayLoad.getExtents();
+    return {extents.begin(), extents.end()};
+  }
+
+  /// Generate the shape of the array expressions based on the destination and
+  /// operand array loads, or the optional evaluate::Shape.
+  llvm::SmallVector<mlir::Value> genIterationShape() {
+    // Use the optional evaluate::Shape if it has constant expressions,
+    // otherwise, use the runtime destination or operand shapes. This could be
+    // finned to find the shape that is the most expressive (for instance, the
+    // one with the most constant extents).
+    auto loc = getLoc();
+    auto destShapeIsConstant = [&]() -> bool {
+      for (const auto &s : *destShape)
+        if (!s || !Fortran::evaluate::IsConstantExpr(*s))
+          return false;
+      return true;
+    };
+    if (destShape.has_value() && destShapeIsConstant()) {
+      auto idxTy = builder.getIndexType();
+      llvm::SmallVector<mlir::Value> shape;
+      for (const auto &s : *destShape)
+        shape.emplace_back(builder.createConvert(
+            loc, idxTy, convertOptExtentExpr(converter, stmtCtx, s)));
+      return shape;
+    }
+    if (destination)
+      return getShape(destination);
+    if (!arrayOperandLoads.empty())
+      return getShape(arrayOperandLoads[0]);
```

Comment:
Maybe consider a stronger guarantee than the first one is the destination array here. Perhaps `llvm::Optional`?

Reply:
I am not sure I understand. What we are taking here is the first array operand, and we are using its shape for the destination. So, `arrayOperandLoads[0]` is not the destination, the destination is created based on this shape.